### PR TITLE
Update update-manager.php

### PR DESCRIPTION
**Pull request by AliasQli**
Update update-manager.php to support advanced grimoire modification. Now it can read files in code/grim/ to make advanced changes to grimoire. Also, you can now use localised umi_ftr.txt instead of the default one.
While making these changes, I maintained its original functions so the generating of English and Russian script won't be influenced.
Below is a guidance on how to use its new features:
> To modify not only the grimoire text but also where they appear in the game, instead of changing the shared files in **code/main**, you can create a **\<locale>_grim.txt** (e.g. cn_grim.txt, ru_grim.txt. \<locale> refers to your location) in **code/grim** to make changes only to your own language. The file should be made up of the sentences glossed by grimoire (e.g. \`"............Kinzo‐{c:86EF9C:san}.\`[gstg 1]), each taking up a line. Then, the new generating program will automatically highlight and insert grimoire entries while taking out the deleted entries. But, if the new generating program finds no such file there, it will fall back onto the default generating mode. Empty lines and comments started by ";" are supported. The **code/grim/cn_grim.txt** is always a good example.
> And, if the number of the grimoire entries in any episode is changed, you'll have to change **script/umi_ftr.txt** too, which is also a shared file. So, to make your modification, you can now create a **umi_ftr.txt** in **script/\<locale>** and make changes to it. Similarly, if the new generating program finds no such file there, it will use the default **umi_ftr.txt**.

Before I solved the problem using the same method but with a js script. This time I rewrote it in php so we can still use the same files that contain the grimoire changes, which will be convenient for us. So I think my changes to the update-manager.php will be more acceptable than yours (whether you have made them or not). After all, it is us that need these functions and we ourselves know us best.
As for omake, we haven't made changes to that. I remember you've said it's easy for you to localise omake so I want to leave the work to you.

### CODE REVIEW

**Comment by vit9696** [comment]
I think we can live with it, but with ftr I believe we should factor out the common part rather than replace the whole part. I will make a closer look closer to the weekend.

**Comment by AliasQli** [comment]
> I think we can live with it, but with ftr I believe we should factor out the common part rather than replace the whole part. I will make a closer look closer to the weekend.

Yes. Most belongs to the common part. The part that needs to be modified is the two functions, `r_grim_display_new` and `r_grim_button_new`, and `dim ?grim[15][5]	;Grimoire array declaration`, I think. Maybe there are more.

**Comment by vit9696** [comment]
I checked umi_ftr.txt changes, and I think it can be merged in master without splitting the languages. From what I remember, that grimoire code is generic, and adding handling code for more values will not break Russian or English anyhow, as they simply do not use these values. It will just be dead code, which we do not care much about.

I noticed that you removed `if %r_side_change = 1 _csp 2 : print 10,100 : mov %r_side_change,0`. Was that intentional and was there any reason for it? I have a commit referencing this change, so I am pretty sure we changed that part intentionally:

```
commit e0c408bb1c37d277b672b2d32717681a224fd7d2
Author: vit9696
Date:   Sun Sep 1 13:03:23 2019 +0000

    Fix potential black screen in grim

diff --git a/scripting/script/umi_ftr.txt b/scripting/script/umi_ftr.txt
index 0b2d05c8..f9e3b29e 100644
--- a/scripting/script/umi_ftr.txt
+++ b/scripting/script/umi_ftr.txt
@@ -37515,7 +37515,11 @@ return
        mov %r_display,2
 
        ;If you selected the tip and there is no flag, put 0 to skip to none selected.
-       if ?grim[%r_display_grim][grim_flg] = 0 mov %r_display_grim,0 : return
+       if ?grim[%r_display_grim][grim_flg] != 0 jumpf
+               mov %r_display_grim,0
+               if %r_side_change = 1 _csp 2 : print 10,100 : mov %r_side_change,0
+               return
+       ~
 
        if %scenario_Number != 1 jumpf ;ep1
        notif %r_display_grim = 1 goto *grim_1
```

Which was a follow-up to the following commit:

```
commit b0f1c5d6156d4304bd6229e188736b1fa8b653bd
Author: vit9696
Date:   Sun Sep 1 12:59:39 2019 +0000

    Fix black screen in tips in ep3

diff --git a/scripting/script/umi_ftr.txt b/scripting/script/umi_ftr.txt
index d0b94f74..0b2d05c8 100644
--- a/scripting/script/umi_ftr.txt
+++ b/scripting/script/umi_ftr.txt
@@ -37282,8 +37282,12 @@ return
        mov %r_display,1
 
        ;If you selected the tip and there is no flag, put 0 to skip to none selected.
-       if ?tips[%r_display_tips][tips_flg] = 0 mov %r_display_tips,0 : return
-       
+       if ?tips[%r_display_tips][tips_flg] != 0 jumpf
+               mov %r_display_tips,0
+               if %r_side_change = 1 _csp 2 : print 10,100 : mov %r_side_change,0
+               return
+       ~
+
        mov %tmp,r_top_y : add %tmp,820
 
        ;Inscription of portrait
```

**Comment by AliasQli** [comment]
Yes you're right. Just set the capacity larger and it will do. I think 20 per episode is enough.
The removal is a careless mistake. Sorry to have bothered you. Please keep it unchanged.

**Comment by AliasQli** [comment]
Yes you're right. Just set the capacity larger and it will do. I think 20 per episode is enough.
The removal is a careless mistake. Sorry to have bothered you. Please keep it unchanged.

**Comment by AliasQli** [comment]
It suddenly occurred to me that your onscripter has difficulty displaying 20 grimoire entries in one episode. Some won't be displayed or will be unable to click. We now have 19 entries in our ep1 and we're facing this problem. We're trying to merge some of the entries but I think a better solution is to add a scroller to your onscripter.

**Comment by AliasQli** [comment]
Ok. Actually neither merging entries nor adding a scroll is my job. If we do have the need, I think other members in our group will talk to you.

**Comment by AliasQli** [comment]
Ok. Actually neither merging entries nor adding a scroll is my job. If we do have the need, I think other members in our group will talk to you.

**Comment by AliasQli** [comment]
Now the change of ftr is restored.

File: update-manager/update-manager.php [hunk 1]
```diff
@@ -460,15 +466,32 @@ function main($argc, $argv) {
...
+			$tmp_script = '';
+			 
 			for ($i = 1; $i <= 8; $i++) {
 				$tldir = $scripting.'/story/ep'.$i.'/'.$locale.'/';
 				if (!is_dir($tldir))
 					$tldir = $scripting.'/story/ep'.$i.'/en/';
-				$script .= inplaceLines($scripting.'/game/main/', $scripting.'/story/ep'.$i.'/jp/', $tldir);
+				$tmp_script .= inplaceLines($scripting.'/game/main/', $scripting.'/story/ep'.$i.'/jp/', $tldir);
+			}
+			$tmp_script .= inplaceLines($scripting.'/game/omake/', $scripting.'/story/omake/jp/',
```
Comment (vit9696) [inline update-manager/update-manager.php]: Omake cannot have grimoire, as it does not have right click menu at all.

File: update-manager/update-manager.php [hunk 2]
```diff
@@ -460,15 +466,32 @@ function main($argc, $argv) {
...
+				$scripting.'/story/omake/'.$locale.'/'); // Maybe omake needs grimoire
+
+			if(file_exists($scripting.'/game/grim/'.$locale.'_grim.txt')) {
+				$grim = preg_replace('/^\s*(;.*)?$/', '', 
+					preg_replace('/\n|\r\n$/', '', file($scripting.'/game/grim/'.$locale.'_grim.txt')) // '/(\r)?\n$/' may not work
+				);
+				$len = count($grim);
+				$preg_grim = [];
+				for ($i = 0; $i < $len; $i++)//{
```
Comment (vit9696) [inline update-manager/update-manager.php]: Maybe kill these comments? At least the `{` / `}` ones.

File: update-manager/update-manager.php [hunk 3]
```diff
@@ -460,15 +466,32 @@ function main($argc, $argv) {
...
+					$preg_grim[$i] = '/'.preg_quote(removeGrim($grim[$i]), '/').'/';
+					//echo "$i#$grim[$i]#$preg_grim[$i]\n";
+				//}
+				$tmp_script = preg_replace($preg_grim, $grim, removeGrim($tmp_script), 1); // Here I only replace it once, because if there are two same sentences, the one that needs grimoire is definitely the first one
```
Comment (vit9696) [inline update-manager/update-manager.php]: Put the comment on the line above, it is not really readable otherwise.

**Comment by AliasQli** [comment]
The three little problems you have mentioned are fixed. As for the gstc update, to be honest I have no idea about it. I haven't dug deep into the grimoire mechanism... Sorry, but I really don't know how to solve the problem.
The mission of this pull request is accomplished. I think it can be merged now.

**Comment by vit9696** [comment]
Hmmm, well, I guess this particular part can indeed be considered as finished, but before we merge this, I should ask you whether you do understand that adding grimoires through this mechanism in places different from Russian/English will break the game?

I.e. they will be opened/closed at times different from their specification if you use use "Chapter select" or "Jump to line" mechanisms at the very least.

**Comment by AliasQli** [comment]
Then... the merge must wait. I didn't know there're so many problems here.

**Comment by copyliu** [comment]
Can we keep the php script not modified and set all EPs grimoire to 10 entrys? then all localization can fill the grimoire whatever they want, just not over 10 entrys. If it don't need 10, it can fill them with spaces.

**Comment by vit9696** [comment]
@copyliu yeah, this part is fine like I said above. The only problem with grimoire is state logic. Update manager changes suggested here are fine too, but they only resolve half of the problem: issues with gstc/gstt compatibility stay. Feel free to submit patches, for some reason the code responsible for increasing the amount of grimoire entries was removed from this PR.

**Comment by copyliu** [comment]
replaced by #8, close

**Pull request closed without merging**